receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream/unmarshaler.go

```diff
@@ -39,61 +45,140 @@ func NewUnmarshaler(logger *zap.Logger) *Unmarshaler {
-		for datumIndex, datum := range bytes.Split(record, []byte(recordDelimiter)) {
-			if len(datum) > 0 {
-				var metric cWMetric
-				err := json.Unmarshal(datum, &metric)
-				if err != nil {
-					u.logger.Error(
-						"Unable to unmarshal input",
-						zap.Error(err),
-						zap.Int("datum_index", datumIndex),
-						zap.Int("record_index", recordIndex),
-					)
-					continue
-				}
-				if !u.isValid(metric) {
-					u.logger.Error(
-						"Invalid metric",
-						zap.Int("datum_index", datumIndex),
-						zap.Int("record_index", recordIndex),
-					)
-					continue
-				}
-				attrs := resourceAttributes{
-					metricStreamName: metric.MetricStreamName,
-					namespace:        metric.Namespace,
-					accountID:        metric.AccountID,
-					region:           metric.Region,
-				}
-				mb, ok := builders[attrs]
-				if !ok {
-					mb = newResourceMetricsBuilder(md, attrs)
-					builders[attrs] = mb
-				}
-				mb.AddMetric(metric)
-			}
-		}
+func (u Unmarshaler) UnmarshalMetrics(record []byte) (pmetric.Metrics, error) {
+	type metricKey struct {
+		name string
+		unit string
 	}
+	byResource := make(map[resourceKey]map[metricKey]pmetric.Metric)
+
+	// Multiple metrics in each record separated by newline character
+	scanner := bufio.NewScanner(bytes.NewReader(record))
+	for datumIndex := 0; scanner.Scan(); datumIndex++ {
+		var cwMetric cWMetric
+		if err := jsoniter.ConfigFastest.Unmarshal(scanner.Bytes(), &cwMetric); err != nil {
+			u.logger.Error(
+				"Unable to unmarshal input",
+				zap.Error(err),
+				zap.Int("datum_index", datumIndex),
+			)
+			continue
+		}
+		if !u.isValid(cwMetric) {
+			u.logger.Error(
+				"Invalid metric",
+				zap.Int("datum_index", datumIndex),
+			)
+			continue
+		}
+
+		rkey := resourceKey{
+			metricStreamName: cwMetric.MetricStreamName,
+			namespace:        cwMetric.Namespace,
+			accountID:        cwMetric.AccountID,
+			region:           cwMetric.Region,
+		}
+		metrics, ok := byResource[rkey]
+		if !ok {
+			metrics = make(map[metricKey]pmetric.Metric)
+			byResource[rkey] = metrics
+		}
 
-	if len(builders) == 0 {
-		return pmetric.NewMetrics(), errInvalidRecords
+		mkey := metricKey{
+			name: cwMetric.MetricName,
+			unit: cwMetric.Unit,
+		}
+		metric, ok := metrics[mkey]
+		if !ok {
+			metric = pmetric.NewMetric()
+			metric.SetName(mkey.name)
+			metric.SetUnit(mkey.unit)
+			metric.SetEmptySummary()
+			metrics[mkey] = metric
+		}
+
+		dp := metric.Summary().DataPoints().AppendEmpty()
+		dp.SetTimestamp(pcommon.NewTimestampFromTime(time.UnixMilli(cwMetric.Timestamp)))
+		setDataPointAttributes(cwMetric, dp)
+		dp.SetCount(uint64(cwMetric.Value.Count))
+		dp.SetSum(cwMetric.Value.Sum)
+		minQ := dp.QuantileValues().AppendEmpty()
+		minQ.SetQuantile(0)
+		minQ.SetValue(cwMetric.Value.Min)
+		maxQ := dp.QuantileValues().AppendEmpty()
+		maxQ.SetQuantile(1)
+		maxQ.SetValue(cwMetric.Value.Max)
+	}
+	if err := scanner.Err(); err != nil {
+		return pmetric.Metrics{}, err
```

Comment:
Can the metrics processed before the error be returned?

Reply:
Will do, same as above